crates/sui-transaction-builder/src/lib.rs

```diff
@@ -0,0 +1,1008 @@
+        #[serde(deserialize_with = "deps_from_str")]
+        dependencies: Vec<ObjectId>,
+        digest: Vec<u8>,
+    }
+
+    fn bcs_from_str<'de, D>(deserializer: D) -> Result<Vec<Vec<u8>>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let bcs = Vec::<String>::deserialize(deserializer)?;
+        bcs.into_iter()
+            .map(|s| base64ct::Base64::decode_vec(&s).map_err(de::Error::custom))
+            .collect()
+    }
+
+    fn deps_from_str<'de, D>(deserializer: D) -> Result<Vec<ObjectId>, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        let deps = Vec::<String>::deserialize(deserializer)?;
+        deps.into_iter()
+            .map(|s| ObjectId::from_str(&s).map_err(de::Error::custom))
+            .collect()
+    }
+
+    /// This is used to read the json file that contains the modules/deps/digest generated with sui
+    /// move build --dump-bytecode-as-base64 on the `test_example_v1 and test_example_v2` projects
+    /// in the tests directory.
+    /// The json files are generated automatically when running `make test-with-localnet` in the
+    /// root of the sui-transaction-builder crate.
+    fn move_package_data(file: &str) -> MovePackageData {
+        let data = std::fs::read_to_string(file)
+            .with_context(|| {
+                format!(
+                    "Failed to read {file}. \
+                    Run `make test-with-localnet` from the root of the repository that will \
+                    generate the right json files with the package data and then run the tests."
+                )
+            })
+            .unwrap();
+        serde_json::from_str(&data).unwrap()
+    }
+
+    /// Generate a random private key and its corresponding address
+    fn helper_address_pk() -> (Address, Ed25519PrivateKey) {
+        let pk = Ed25519PrivateKey::generate(rand::thread_rng());
+        let address = pk.public_key().to_address();
+        (address, pk)
+    }
+
+    /// Generate a private key and its corresponding address, call faucet which returns 5 coin
+    /// objects, set the sender to this address, set the gas object (last coin from the returned
+    /// Vec<CoinInfo>), set gas price, set gas budget, and return the address, private key, and
+    /// coins.
+    ///
+    /// NB! This assumes that these tests run on a network whose faucet returns 5 coins per
+    /// each faucet request.
+    async fn helper_setup(
+        tx: &mut TransactionBuilder,
+        client: &Client,
+    ) -> (Address, Ed25519PrivateKey, Vec<CoinInfo>) {
+        let (address, pk) = helper_address_pk();
+        let coins = FaucetClient::local()
+            .request_and_wait(address)
+            .await
+            .unwrap()
+            .unwrap()
+            .sent;
+        println!("Coins: {:?}", coins);
+        let gas = coins.last().unwrap().id;
+        let gas_obj = client.object(gas.into(), None).await.unwrap().unwrap();
+        tx.add_gas(vec![gas_obj]);
```

Comment:
In theory yes that should work. In practice, now, it will fail because it requires to figure out the version and digest of an object and we don't have the transaction resolution plugged in yet.

Reply:
After fixing `IntoInput` impl, passing `&gas_obj` will not work either because the object has an `Address` owner and we do not know if it's owned or receiving type, so we have to build the object by passing the version and digest in addition to the object id.

Reply:
In this case, maybe it's worth having a special function on `Object` to convert it into an `Input` that you can then modify (see the discussion in my latest round of review comments on the `unresolved::Input` builder APIs):

```rust
impl Object {
    pub fn as_input(&self) -> unresolved::Input;
}

tx.add_gas([gas_obj.as_input().with_owned_kind()]);
```

Reply:
We'd still need to get the object because we need version + digest. Object itself does not have a digest, so we'd have to compute it / add a function on Object to compute it.

Reply:
Sorry, Object does have a digest that can be computed already. Hmm.

Reply:
Added it!